manuals/fsrefsinfo.1
.Dd June 26, 2026
.Dt FSREFSINFO 1
.Os
.Sh NAME
.Nm fsrefsinfo
.Nd determines information about a Resiliant File System (ReFS) volume
.Sh SYNOPSIS
.Nm fsrefsinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm fsrefsinfo
is a utility to determine information about a Resiliant File System (ReFS) \
volume
.Pp
.Nm fsrefsinfo
is part of the
.Nm libfsrefs
package.
.Nm libfsrefs
is a library to access the Resiliant File System (ReFS) format
.Pp
.Ar source
is the source volume.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fsrefsinfo /dev/sda1
fsrefsinfo 20260626
.sp
Resiliant File System (ReFS) volume information:
.sp
Volume information:
	Name				: Test Volume
	Version				: 1.1
	Serial number			: bf790b1be6690b15
	Bytes per sector		: 512
	Cluster block size		: 65536
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfsrefs/issues
.Sh COPYRIGHT
Copyright (C) 2012-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO

Format support
* Use checkpoint in determining block offsets
* Rename metadata block header to block header ?
* Rename metadata block size to page size ?
* Use directory object name records in determining directory entries and lookup detailed entry record on demand
* Add support for multi node file values ministore tree

API
* add function to retrieve container size

20141011
* see `git log' for more recent change log
* removed README.macosx
* updated dependencies
* changes for project site move

20140529
* updated dependencies

20140118
* 2014 update
* updated dependencies

20131109
* updated dependencies
* code clean up

20130915
* code clean up

20130914
* updated dependencies
* added libcthreads build support

20130224
* updated dependencies
* 2013 update

20120930
* worked on format

20120929
* updated dependencies
* worked on format

20120528
* worked on format

20120527
* initial version based on libfsntfs



fsrefstools/fsrefsinfo.c
/*
 * Shows information obtained from a Resiliant File System (ReFS) volume.
 *
 * Copyright (C) 2012-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fsrefstools_getopt.h"
#include "fsrefstools_libcerror.h"
#include "fsrefstools_libclocale.h"
#include "fsrefstools_libcnotify.h"
#include "fsrefstools_libfsrefs.h"
#include "fsrefstools_output.h"
#include "fsrefstools_signal.h"
#include "fsrefstools_unused.h"
#include "info_handle.h"

enum FSREFSINFO_MODES
{
	FSREFSINFO_MODE_FILE_ENTRY,
	FSREFSINFO_MODE_FILE_SYSTEM_HIERARCHY,
	FSREFSINFO_MODE_USN_CHANGE_JOURNAL,
	FSREFSINFO_MODE_VOLUME
};

info_handle_t *fsrefsinfo_info_handle = NULL;
int fsrefsinfo_abort                  = 0;

/* Signal handler for fsrefsinfo
 */
void fsrefsinfo_signal_handler(
      fsrefstools_signal_t signal FSREFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fsrefsinfo_signal_handler";

	FSREFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fsrefsinfo_abort = 1;

	if( fsrefsinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     fsrefsinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fsrefsinfo to determine information about a Resiliant File System (ReFS) volume.";

	fsrefstools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'H', NULL, "shows the file system hierarchy" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source volume" },
	};
	system_character_t options_string[ 32 ];

	libfsrefs_error_t *error                 = NULL;
	system_character_t *option_volume_offset = NULL;
	system_character_t *source               = NULL;
	char *program                            = "fsrefsinfo";
	system_integer_t option                  = 0;
	uint8_t calculate_md5                    = 0;
	int number_of_options                    = (int) ( sizeof( options ) / sizeof( fsrefstools_option_t ) );
	int option_mode                          = FSREFSINFO_MODE_VOLUME;
	int verbose                              = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fsrefstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fsrefstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fsrefstools_output_version_fprint(
	 stdout,
	 program );

	if( fsrefstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fsrefstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fsrefstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				fsrefstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'H':
				option_mode = FSREFSINFO_MODE_FILE_SYSTEM_HIERARCHY;

				break;

			case (system_integer_t) 'o':
				option_volume_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fsrefstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		fsrefstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libfsrefs_notify_set_stream(
	 stderr,
	 NULL );
	libfsrefs_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &fsrefsinfo_info_handle,
	     calculate_md5,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( fsrefsinfo_info_handle != NULL );
#endif
	if( option_volume_offset != NULL )
	{
		if( info_handle_set_volume_offset(
		     fsrefsinfo_info_handle,
		     option_volume_offset,
		     &error ) != 1 )
		{
			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );

			fprintf(
			 stderr,
			 "Unsupported volume offset defaulting to: %" PRIi64 ".\n",
			 fsrefsinfo_info_handle->volume_offset );
		}
	}
	if( info_handle_open_input(
	     fsrefsinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	switch( option_mode )
	{
		case FSREFSINFO_MODE_FILE_SYSTEM_HIERARCHY:
			if( info_handle_file_system_hierarchy_fprint(
			     fsrefsinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print file system hierarchy.\n" );

				goto on_error;
			}
			break;

		case FSREFSINFO_MODE_VOLUME:
		default:
			if( info_handle_volume_fprint(
			     fsrefsinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print volume information.\n" );

				goto on_error;
			}
			break;
	}
	if( info_handle_close_input(
	     fsrefsinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &fsrefsinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( fsrefsinfo_info_handle != NULL )
	{
		info_handle_free(
		 &fsrefsinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

